File: app/src/modules/clients/api/services/client-service.ts
import { Client } from '@/modules/clients/api/entities/client';
import { CreateClient } from '@/modules/clients/api/data-transfer-objects/create-client';
import { UpdateClient } from '@/modules/clients/api/data-transfer-objects/update-client';
import { http } from '@/core';

export class ClientService {
    async getClients(): Promise<Client[]> {
        const res = await http.get<any[]>('/client');
        return res.data.map(r => this._map(r));
    }

    async createClient(create: CreateClient): Promise<Client> {
        const res = await http.post('/client', create);
        return this._map(res.data);
    }

    async updateClient(update: UpdateClient): Promise<Client> {
        const res = await http.put(`/client/${update.id}`, update);
        return this._map(res.data);
    }

    async deleteClient(client: Client): Promise<void> {
        await http.delete(`client/${client.id}`);
    }

    _map(a: any): Client {
        var c = new Client(a.name, a.phone, a.email);
        c.id = a.id;

        return c;
    }
}

File: api/src/Domain/Clients/Services/ClientService.cs
using System;
using System.Collections.Generic;
using System.Threading.Tasks;

namespace DetailingArsenal.Domain.Clients {
    public interface IClientService : IService {
        Task<Client> GetById(Guid id);
        Task<List<Client>> GetByUser(User user);
        Task<Client> Create(CreateClient create, User user);
        Task Update(Client client, UpdateClient update);
        Task Delete(Client client);
    }

    public class ClientService : IClientService {
        IClientRepo repo;
        ClientHasNoAppointmentsSpecification noAppointmentsSpec;

        public ClientService(IClientRepo repo, ClientHasNoAppointmentsSpecification noAppointmentsSpec) {
            this.repo = repo;
            this.noAppointmentsSpec = noAppointmentsSpec;
        }

        public async Task<Client> GetById(Guid id) {
            return await repo.FindById(id) ?? throw new EntityNotFoundException();
        }

        public async Task<List<Client>> GetByUser(User user) {
            return await repo.FindByUser(user);
        }

        public async Task<Client> Create(CreateClient create, User user) {
            var c = Client.Create(
                user.Id,
                create.Name,
                create.Phone,
                create.Email
            );

            await repo.Add(c);
            return c;
        }

        public async Task Update(Client client, UpdateClient update) {
            client.Name = update.Name;
            client.Phone = update.Phone;
            client.Email = update.Phone;

            await repo.Update(client);
        }

        public async Task Delete(Client client) {
            await noAppointmentsSpec.CheckAndThrow(client);
            await repo.Delete(client);
        }
    }
}